core/src/main/java/org/apache/iceberg/SnapshotOperations.java

```diff
@@ -0,0 +1,202 @@
+  private Map<String, SnapshotRef> refs;
+
+  SnapshotOperations() {
+    snapshots = Lists.newArrayList();
+    snapshotsById = Maps.newHashMap();
+    refs = Maps.newHashMap();
+  }
+
+  SnapshotOperations(List<Snapshot> snapshots, Map<String, SnapshotRef> refs) {
+    this(snapshots, null, refs);
+  }
+
+  SnapshotOperations(SnapshotOperations fromCopy) {
+    this(fromCopy.snapshots, fromCopy.snapshotsSupplier, fromCopy.refs);
+  }
+
+  SnapshotOperations(
+      List<Snapshot> snapshots,
+      Supplier<List<Snapshot>> snapshotsSupplier,
+      Map<String, SnapshotRef> refs) {
+    this.snapshots = Lists.newArrayList(snapshots);
+    this.snapshotsSupplier = snapshotsSupplier;
+    this.refs = Maps.newHashMap(refs);
+
+    this.snapshotsById = indexSnapshotsById(snapshots);
+  }
+
+  List<Snapshot> snapshots() {
+    ensureLoaded();
+
+    return Collections.unmodifiableList(snapshots);
+  }
+
+  Snapshot snapshot(long id) {
+    if (!snapshotsById.containsKey(id)) {
+      ensureLoaded();
+    }
+
+    return snapshotsById.get(id);
+  }
+
+  boolean contains(long id) {
+    if (!snapshotsById.containsKey(id)) {
+      ensureLoaded();
+    }
+
+    return snapshotsById.containsKey(id);
+  }
+
+  void add(Snapshot snapshot) {
+    ValidationException.check(
+        !contains(snapshot.snapshotId()),
+        "Snapshot already exists for id: %s",
+        snapshot.snapshotId());
+
+    snapshots.add(snapshot);
+    snapshotsById.put(snapshot.snapshotId(), snapshot);
+  }
+
+  List<Snapshot> remove(Collection<Long> idsToRemove) {
+    List<Snapshot> retainedSnapshots =
+        Lists.newArrayListWithExpectedSize(snapshots.size() - idsToRemove.size());
+    List<Snapshot> removedSnapshots = Lists.newArrayListWithExpectedSize(idsToRemove.size());
+
+    for (Snapshot snapshot : snapshots) {
+      long snapshotId = snapshot.snapshotId();
+      if (idsToRemove.contains(snapshotId)) {
+        Snapshot removed = snapshotsById.remove(snapshotId);
+        removedSnapshots.add(removed);
+      } else {
+        retainedSnapshots.add(snapshot);
+      }
+    }
+
+    this.snapshots = retainedSnapshots;
+
+    // remove any refs that are no longer valid
+    Set<String> danglingRefs = Sets.newHashSet();
+    for (Map.Entry<String, SnapshotRef> refEntry : refs.entrySet()) {
+      if (!snapshotsById.containsKey(refEntry.getValue().snapshotId())) {
+        danglingRefs.add(refEntry.getKey());
+      }
+    }
+
+    danglingRefs.forEach(this::removeRef);
+
+    return removedSnapshots;
+  }
+
+  Map<String, SnapshotRef> refs() {
+    return Collections.unmodifiableMap(refs);
+  }
+
+  SnapshotRef ref(String name) {
+    return refs.get(name);
+  }
+
+  void addRef(String name, SnapshotRef ref) {
+    refs.put(name, ref);
+  }
+
+  SnapshotRef removeRef(String name) {
+    return refs.remove(name);
+  }
+
+  private void ensureLoaded() {
+    if (snapshotsSupplier != null) {
+      snapshots.addAll(snapshotsSupplier.get());
```

Comment:
This can potentially reorder snapshots, which would be a problem for checks in the `TableMetadata` constructor. I think we should revisit that check, since branching probably makes it obsolete.

Reply:
I've updated this because the snapshots should be replaced entirely.